### ISSUE TITLE
Validating entire form instead of a single page (AJAX)

### ISSUE THREAD

**Issue by joshuabaker** [verbb/formie#618]
## Description

We’re encountering a bug. When submitting a page by AJAX the validation rules for subsequent pages, instead of just the `pageIndex` provided.

**Form Data**

```
handle: conditional
pageIndex: 0
fields[showHiddenField]: No
fields[hiddenField]: 
fields[page2Required]: 
fields[page3Required]: 
```

**Response**

```js
{
  "success": false,
  "submissionId": null,
  "currentPageId": "1238",
  "nextPageId": null,
  "nextPageIndex": 2,
  "totalPages": 3,
  "redirectUrl": "",
  "submitActionMessage": "Submission saved.",
  "errors":
  {
    "page2Required":
    [
      "Page 2 Required cannot be blank."
    ],
    "page3Required":
    [
      "Page 3 Required cannot be blank."
    ]
  },
  "errorMessage": "Couldn’t save submission due to errors."
}
```

In this case, I’d expect only the validation schema for the selected page would be applied, which I’m almost certain has been the case until a recent version. Neither `page2Required` nor `page3Required` are present on the first page.

## Form settings
- Multi-page form: Yes
- Submission Method: Ajax
- Client-side Validation: Yes
- Custom Form Templates: N/a

## Additional info
- Plugin version: 1.4.22
- Craft version: 3.7.17.2
- Multi-site: Yes or No

## Additional context

We’re adding conditional support to [`react-formie-forms`](https://www.npmjs.com/package/react-formie-forms).

I’ve tried mocking this using Insomnia with the same result.

**Comment by joshuabaker** [comment]
PHP 8 evaluates `'0'` as `false`, which means that a `pageIndex` of zero skips the condition I added in 1.3.11.

**Comment by engram-design** [comment]
I don't believe we use the `pageIndex` param, so will have to be guided by you on this (https://github.com/verbb/formie/pull/306).

Not sure if it helps in your case, but you can also use the `completeSubmission` if you want to submit the payload for the entire form (if your React app handles the validation on the front-end, and you just want to submit it all in one go).

**Comment by joshuabaker** [comment]
`pageIndex` appears in form templates (below).

https://github.com/verbb/formie/blob/98fe24d54805d21f7bd0e99cdd69853c052d93b6/src/templates/_special/form-template/form.html#L72

Either way, it’s impossible to validate partial submissions via AJAX without `pageIndex` so this needs to be supported properly.

I note that the logic (below) in the [SubmissionsController](https://github.com/verbb/formie/blob/98fe24d54805d21f7bd0e99cdd69853c052d93b6/src/controllers/SubmissionsController.php) has been updated without consideration for the `pageIndex` param, effectively breaking its function. #619 fixed the PHP 8 compatibility of the condition, however, [`$currentPage`](https://github.com/verbb/formie/blob/98fe24d54805d21f7bd0e99cdd69853c052d93b6/src/controllers/SubmissionsController.php#L372) is never actually used anymore! I think you need to add some unit tests to prevent changes like this from just breaking features.

https://github.com/verbb/formie/blob/98fe24d54805d21f7bd0e99cdd69853c052d93b6/src/controllers/SubmissionsController.php#L391-L401

Under the hood, `completeSubmission` is simply setting the `pageIndex` to the last page (i.e. `"pageIndex": pages.length - 1`). It’s arguably redundant and contends with the `pageIndex` param. You’ll need to validate params now, I think, to prevent unexpected behaviour.

**Comment by engram-design** [comment]
You're correct that `pageIndex` exists in the templates, but we don't really use it for much (as shown by that conditional block doing nothing).

https://github.com/verbb/formie/blob/98fe24d54805d21f7bd0e99cdd69853c052d93b6/src/controllers/SubmissionsController.php#L391-L394

Looking at your initial PR (where you added the `$currentPage` variable and the `pageIndex` param), it's never done anything? https://github.com/verbb/formie/pull/306/commits/479db8d34522d03228234d0d681d4bee0516ea6e

I've just pushed a commit to clean this up and ensure it's properly setting (overriding) the current page. https://github.com/verbb/formie/commit/9e1cce8408cf3b1aad0ffbbde220452bff82f8a7

My https://github.com/verbb/formie/issues/319#issuecomment-767467165 was going to change the behaviour of this by adding a `pageIndex` param in templates, rather than relying on the session to determine the current page (to therefore validate), but it'll break too many people's custom templates where param won't exist. As such, `pageIndex` shouldn't exist at the moment in the templates at all, and looks like I've added that in by mistake (likely during my testing, where I deemed this to be a breaking change in https://github.com/verbb/formie/commit/b8a3cd5d403b4b9f012d645efbbacba204dc6d1a). We'll change this at some point so it's consistent across all scenarios for sanity.

In a headless context, or where you're running your own templates, I agree that you should be able to control the current page through the `pageIndex` param. Formie's controller will try and fetch the current page and validate just that, but in a headless scenario, there's no session to maintain, so requiring this param to be provided is for that.

`completeSubmission` might be a tiny bit overkill (like you mention), but it's also for users who don't maintain a copy of the number of pages client-side, and would be cumbersome to do `pages.length - 1`. As such, it is intended to override `pageIndex` and force the last page as the current page. https://github.com/verbb/formie/issues/319

**Comment by joshuabaker** [comment]
Ah, amazing. I was just working up a PR that does nearly exactly what you’ve done in 9e1cce8408cf3b1aad0ffbbde220452bff82f8a7.

Thanks for the clarification on those points. Tend to agree that `completeSubmission` is a handy shortcut.

**Comment by joshuabaker** [comment]
I’ve encountered a further issue here that’s related to page conditions.

It’s currently possible to set page conditions on fields that have not yet been encountered. If a user implements this based on a field from the last page in a form (e.g. selecting “Show page 2” on page 3 of 3) partial submissions will fail as the preceding page would fail validation.

![local-dandaraliving_8888_cms_formie_forms_edit_21934(iPad Pro)](https://user-images.githubusercontent.com/160484/138951293-d95a323b-fb7f-4937-b469-258cda399139.png)

Is there any way that we can prevent full submission validation on the last page to counter this?

**Comment by joshuabaker** [comment]
Ideally, page conditions available would be limited to only those currently encountered (i.e. fields from page 1 on page 2, fields from page 1 and page 2 on page 3, and so on). This would prevent form editors from selecting options that they probably shouldn’t.

I can see that this is a breaking change.

**Comment by engram-design** [comment]
Hmm, in short not really. It's one of the issues I faced when implementing multi-page conditions (both page reload and ajax), and had to modify the behaviour in templates to submit the entire form payload each time. See https://github.com/verbb/formie/issues/521

The main issue I can see at the moment is that using the Page Tabs function, you can navigate to different pages. No validation occurs on this action, because its navigating, not submitting. This gives users the ability to go back to previous pages, or skip to next pages. As you can guess, through this method you can technically navigate right to the last page of a form, hit submit, and any page before the last page with required fields or validation handling will skip it, because it'll only validate the last page.

As such, validating the entire form is a bit of a safety net to prevent users missing field validations.

One way around this is to deliver the entire payload each time you hit the `submit` action. Still submitting the `pageIndex` value so that it only validates the current page, but conditions are unaffected. This is what the Formie default templates do (without the `pageIndex`), and it does require maintaining a model of the form client-side - if that's possible for your setup.

The other options are as you've mentioned, preventing "future" fields from even being used in page conditions - which probably isn't a bad idea in general, and probably should've had the foresight to see it'd cause issues. I don't _think_ there's a valid scenario where it should be allowed.

We could also disable the full-page submission validation when passing in the `pageIndex` param, but it does require implementations for multi-step forms to change.

**Comment by joshuabaker** [comment]
> One way around this is to deliver the entire payload each time you hit the `submit` action

We are doing this actually, but it won’t work, at present. If the `pageIndex` is the last page [the entire form is validated regardless](https://github.com/verbb/formie/blob/1a9cc1544528b6102f4b6329089b02a4abc13f84/src/controllers/SubmissionsController.php#L444-L445). If any preceding page fails validation the whole request fails (i.e. partial submission cannot be saved).

I would suggest that we prevent ‘future’ fields from being selected, which addresses the underlying cause of these symptoms. This makes sense generally and will prevent a lot of logic complexity now and in future. Anything else is going to be working around this (e.g. adjusting `pageIndex` logic somehow and/or requiring clients to pass `completeSubmission`).

**Comment by engram-design** [comment]
So I suppose my question is, why are proceeding pages failing validation, when the last and final page is being submitted (so all pages are validated)? Each page should be submitted and validated, and once passed, moved onto the next page. It's only at the final page there's the final validation-everything action, but should complete fine, under normal circumstances.

Or, maybe that's what you're getting at in your example. You're on Page 3, and you add a content to a field, which is used in a previous page's conditions (Page 2 is shown when this Page 3 field is true). Therefore, with the Page 2 now visible, and with potential required fields, that'll throw a validation error, because you need to go back and fill in that now-visible page.

So I agree that future-fields fix will sort that. Do you need to address that now in your forms, or will this fix solve that (and for now, just be careful not to create conditions like that)?

**Comment by engram-design** [comment]
As an aside, a recent commit (https://github.com/verbb/formie/commit/043ff8c09799a8fb2dcbcd76211b841b5a109ef5) actually is supposed to redirect to the first page with an error. This is included in the payload response from any call as the `nextPageIndex`. So in your case, on final submit, it should return that `nextPageIndex = 1` if that helps in any way...

**Comment by joshuabaker** [comment]
The ‘redirect to first invalid page’ logic is something that we’ve built in already. It’s that the last page hits the full form validation that’s getting in our way, and this is only because someone can current do weird things with page conditions.

We don’t have this exact case now in a _real_ form. As I put together the changes to [joshuabaker/react-formie-forms](https://github.com/joshuabaker/react-formie-forms), I was keen to test out as many scenarios as made possible in the CP interface and this logic quirk surfaced.

I’ll leave you to consider the approach here and am happy to feedback, if needed, but very much welcome a fix to the ‘future’ fields issue.

**Comment by engram-design** [comment]
Yep agreed, I'll circle back to this shortly!

**Comment by joshuabaker** [comment]
Just a follow up thought; Should it be possible for the last page to be conditional?

**Comment by engram-design** [comment]
Yep, I think a conditional last page is pretty valid! A good use-case is if you don't meet a condition in a previous page, you won't see or reach the last page in the form. Maybe that last page is contact information for a service you haven't picked in a previous page.

In that instance, the last visible page becomes the final page. The controller action will look at the next non-conditionally-hidden page.

**Comment by joshuabaker** [comment]
That’s going to confuse scenarios where you’ve got “Next” as your submit button text; A common button label. Thinking on this one further, and the possible scenarios available getting messy, I’d be inclined to omit the last page from page conditions.

**Comment by engram-design** [comment]
Hmm, another conversation perhaps. We have a couple of clients using it, but yes the button issue is a potential UX concern where you have "Next" on all middle pages, but "Send/Submit" on the final page. Changing that would be tricky (from a content management point of view).

I'll consider it.

**Comment by engram-design** [comment]
So just to revise here (as it's been a while).

> Is there any way that we can prevent full submission validation on the last page to counter this?

This will never be the case unfortunately. This is because users can to back to previous pages, save a page and navigate using the tabs to future pages - all bypassing validation (skipping required fields, etc), so it's vitally important we maintain the final last-step triggers validation for the entire form. It's also a good idea to just not assume every previous page is "done and valid" when the user could be coming back hours/days/weeks to pick the submission up again, when field requirements might've changed.

> Ideally, page conditions available would be limited to only those currently encountered (i.e. fields from page 1 on page 2, fields from page 1 and page 2 on page 3, and so on). This would prevent form editors from selecting options that they probably shouldn’t.

This is probably a good approach. The tricky part is handling when fields are moved, or pages are moved their conditions might get messed up. Say you move the last page with conditions setup for fields on previous pages to the 2nd page, the fields might not be valid. I think that's pretty fair game though.

Fixed for the next release. To get this fix early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-4 as 2.0.0-beta.16",
  "...": "..."
}
```

Then run `composer update`. 

> Just a follow up thought; Should it be possible for the last page to be conditional? That’s going to confuse scenarios where you’ve got “Next” as your submit button text; A common button label. Thinking on this one further, and the possible scenarios available getting messy, I’d be inclined to omit the last page from page conditions.

I don't think there's a good way around this, as we need to allow users to conditionally hide final pages. We have one client that has a 5 page form, but only the first 2 are required for some conditions. As such, the second page becomes the final page, so even if we were to disable conditions on the last page, that wouldn't cater for that.

The only _real_ way to handle that would be to grab the button settings (position, text, JS events) from the last page, if conditionally hidden, and override the last non-conditionally-hidden page's buttons. But that seems like strange behaviour.

As such, going to close this for the moment to cleanup this thread from the original issue.

**Comment by engram-design** [comment]
Fixed in [2.0.0](https://github.com/verbb/formie/releases/tag/2.0.0)
